src/backtracking/hamiltonian_cycle.rs

```diff
@@ -0,0 +1,298 @@
+    /// * `pos` - The position of the current vertex being considered.
+    ///
+    /// # Returns
+    ///
+    /// `true` if it is safe to include `v` in the path, `false` otherwise.
+    fn is_safe(&self, v: usize, visited: &HashSet<usize>, path: &[usize], pos: usize) -> bool {
+        // Check if the current vertex and the last vertex in the path are adjacent.
+        if !self.adjacency_matrix[path[pos - 1]][v] {
+            return false;
+        }
+
+        // Check if the vertex has already been included in the path.
+        !visited.contains(&v)
+    }
+
+    /// Recursively searches for a Hamiltonian cycle.
+    ///
+    /// This function is called by `find_hamiltonian_cycle`.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - A mutable vector representing the current path being explored.
+    /// * `visited` - A mutable set representing the visited vertices.
+    /// * `pos` - The position of the current vertex being considered.
+    ///
+    /// # Returns
+    ///
+    /// `true` if a Hamiltonian cycle is found, `false` otherwise.
+    fn hamiltonian_cycle_util(
+        &self,
+        path: &mut Vec<usize>,
+        visited: &mut HashSet<usize>,
+        pos: usize,
+    ) -> bool {
+        if pos == self.num_vertices() {
+            // Check if there is an edge from the last included vertex to the first vertex.
+            return self.adjacency_matrix[path[pos - 1]][path[0]];
+        }
+
+        for v in 0..self.num_vertices() {
+            if self.is_safe(v, visited, path, pos) {
+                path[pos] = v;
+                visited.insert(v);
+                if self.hamiltonian_cycle_util(path, visited, pos + 1) {
+                    return true;
+                }
+                path[pos] = usize::MAX;
+                visited.remove(&v);
+            }
+        }
+
+        false
+    }
+
+    /// Attempts to find a Hamiltonian cycle in the graph, starting from the specified vertex.
+    ///
+    /// A Hamiltonian cycle visits every vertex exactly once and returns to the starting vertex.
+    ///
+    /// # Note
+    /// This implementation may not find all possible Hamiltonian cycles.
+    /// It stops as soon as it finds one valid cycle. If multiple Hamiltonian cycles exist,
+    /// only one will be returned.
+    ///
+    /// # Returns
+    ///
+    /// `Ok(Some(path))` if a Hamiltonian cycle is found, where `path` is a vector
+    /// containing the indices of vertices in the cycle, starting and ending with the same vertex.
+    ///
+    /// `Ok(None)` if no Hamiltonian cycle exists.
+    fn find_hamiltonian_cycle(
+        &self,
+        start_vertex: usize,
+    ) -> Result<Option<Vec<usize>>, FindHamiltonianCycleError> {
+        // Validate the start vertex.
+        if start_vertex >= self.num_vertices() {
+            return Err(FindHamiltonianCycleError::StartOutOfBound);
+        }
+
+        // Initialize the path.
+        let mut path = vec![usize::MAX; self.adjacency_matrix.len()];
```

Comment:
Thing to consider: why not to represent `path` as `Vec<Option<usize>>`. `None`, would represent the _undefined_ node.

Furthermore:
```suggestion
        let mut path = vec![usize::MAX; self.num_vertices()];
```

Reply:
It's a good suggestion, I will consider it.